storage/range.go

```diff
@@ -1317,15 +1315,30 @@ func (r *Range) InternalMerge(batch engine.Engine, ms *engine.MVCCStats, args *p
 
 // InternalTruncateLog discards a prefix of the raft log.
 func (r *Range) InternalTruncateLog(batch engine.Engine, ms *engine.MVCCStats, args *proto.InternalTruncateLogRequest, reply *proto.InternalTruncateLogResponse) {
+	// args.Index is the first index to keep.
+	term, err := r.Term(args.Index - 1)
+	if err != nil {
+		reply.SetGoError(err)
+		return
+	}
 	start := engine.RaftLogKey(r.Desc().RaftID, args.Index).Next()
 	end := engine.RaftLogKey(r.Desc().RaftID, 0)
-	err := batch.Iterate(engine.MVCCEncodeKey(start), engine.MVCCEncodeKey(end),
+	err = batch.Iterate(engine.MVCCEncodeKey(start), engine.MVCCEncodeKey(end),
 		func(kv proto.RawKeyValue) (bool, error) {
 			err := batch.Clear(kv.Key)
 			return false, err
 		})
+	if err != nil {
+		reply.SetGoError(err)
+		return
+	}
+	ts := proto.RaftTruncatedState{
+		Index: args.Index - 1,
+		Term:  term,
+	}
+	err = engine.MVCCPutProto(batch, nil, engine.RaftTruncatedStateKey(r.Desc().RaftID),
```

Comment:
Looks like InternalTruncateLog must not run concurrently because two concurrent truncates with different `args.Index` could create inconsistent state if the batch with the lower `args.Index` makes it through RocksDB first, with higher entries having been deleted regardless.
This is only true assuming `batch.Clear` doesn't mind clearing out keys which have been cleant before (which I don't think but also don't know).
Maybe this method needs the usual `r.Lock()`?